### ISSUE TITLE
Creating large Endpoints objects in large clusters takes long

### ISSUE THREAD

**Issue by mm4tt** [kubernetes/kubernetes#75294]
**What happened**:
Creating large (e.g. mapping to 5K pods) Endpoints objects takes really long - from a few seconds to over ten seconds. 
We've noticed that in recent runs of our scalability tests (see https://github.com/kubernetes/kubernetes/issues/75284).
Below few examples (all from 5k nodes clusters)

1.  Creating Endpoints object for 5K kube-proxy pods:

    [Run 330](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/330)
    ```
    Trace[866828359]: "Create /api/v1/namespaces/kube-system/endpoints" (started: 2019-03-11 08:20:14.448210695 +0000 UTC m=+579.430813234) (total time: 14.520559394s):
    Trace[866828359]: [165.764543ms] [149.545551ms] Object stored in database
    Trace[866828359]: [14.304157451s] [14.138392908s] Writing response
    Trace[866828359]: [14.520559394s] [216.401943ms] END
    ```

    [Run 328](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/328)
    ```
    Trace[1616289729]: "Create /api/v1/namespaces/kube-system/endpoints" (started: 2019-03-07 08:41:21.676822803 +0000 UTC m=+704.420015010) (total time: 11.451670479s):
    Trace[1616289729]: [152.81293ms] [137.568607ms] Object stored in database
    Trace[1616289729]: [3.354180559s] [3.201367629s] Writing response
    Trace[1616289729]: [11.451670479s] [8.09748992s] END
    ```

2.  Creating Endpoints object for 5K kubelet pods:

    [Run 328](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/328)
    ```
    Trace[2055230634]: "Create /api/v1/namespaces/kube-system/endpoints" (started: 2019-03-07 08:44:50.947415795 +0000 UTC m=+913.690607857) (total time: 4.66432266s):
    Trace[2055230634]: [218.040902ms] [190.653528ms] Object stored in database
    Trace[2055230634]: [3.845597127s] [3.627556225s] Writing response
    Trace[2055230634]: [4.66432266s] [818.725533ms] END
    ```

    [Run 327](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/327)
    ```
    Trace[1293862346]: "Create /api/v1/namespaces/kube-system/endpoints" (started: 2019-03-06 08:46:28.771354922 +0000 UTC m=+1123.112308384) (total time: 5.100958408s):
    Trace[1293862346]: [214.536261ms] [177.914922ms] Object stored in database
    Trace[1293862346]: [1.373099473s] [1.158563212s] Writing response
    Trace[1293862346]: [5.100958408s] [3.727858935s] END
    ```

**What you expected to happen**:
It shouldn't take longer than 1s

**How to reproduce it (as minimally and precisely as possible)**:
Create Endpoints object mapping to 5K addresses.
Alternatively create service mapping to 5K pods (large cluster required).

**Comment by mm4tt** [comment]
I tried to reproduce it on smaller cluster (50 nodes), with artificially big master (n1-standard-64, to match the 5K cluster setup).

I created a big endpoints object (8192 ip addresses) via kubectl apply, but the latency was satisfactory. I repeated the test few times and every time it took ~1s and generated a trace similar to this one:

```
I0312 18:23:35.687919       1 wrap.go:47] POST /api/v1/namespaces/test/endpoints: (1.128965033s) 201 [kubectl/v1.13.4 (linux/amd64) kubernetes/c27b913 104.132.162.86:39527]
Trace[952901299]: "Create /api/v1/namespaces/test/endpoints" (started: 2019-03-12 18:23:34.559418981 +0000 UTC m=+5717.171135292) (total time: 1.118640793s):
Trace[952901299]: [563.524098ms] [563.524098ms] About to convert to expected version
Trace[952901299]: [740.678634ms] [162.759315ms] Object stored in database
Trace[952901299]: [1.118640793s] [343.008595ms] END
```

**Comment by mm4tt** [comment]
I've reserved a scale-testing slot on 03/26 to reproduce the issue on the real 5K node cluster

**Comment by mm4tt** [comment]
One more thing to add. I've analyze the two runs of the scale tests where the issue manifested.

**Run 328**
![AtXx1uk8T5j](https://user-images.githubusercontent.com/2604887/54266135-178c3480-4577-11e9-9186-38ea2f95eb84.png)

**Run 330**
![COB24XApjh6](https://user-images.githubusercontent.com/2604887/54266080-f75c7580-4576-11e9-9891-8f726e55814e.png)

It looks like the issue manifests only in big cluster with high load on API server.
There is nothing surprising happening before or after the endpoints creation requests (QPS, latency, etc. looks pretty normal).

It looks like the endpoints creation requests not only take long, but they also slow down the whole API server.

**Comment by wojtek-t** [comment]
I think we roughly understood that (it doesn't seem to be a regression which is the most important thing). @mm4tt will describe a bit more details soon

**Comment by mm4tt** [comment]
I was able to reproduce it on my "fake large" cluster (50 nodes, n1-standard-64 master).

The difference now is that I started 5000 watches on the Endpoints objects to simulate what kube-proxy does in the real large cluster.
With the watches the request to create a large Endpoints (8K adresses) took 4s and has very similar trace to the real one:
```
I0313 17:16:57.758211       1 trace.go:81] Trace[501784879]: "Create /api/v1/namespaces/test/endpoints" (started: 2019-03-13 17:16:53.707017357 +0000 UTC m=+1113.265325466) (total time: 4.0511125s):
Trace[501784879]: [564.899174ms] [564.899174ms] About to convert to expected version
Trace[501784879]: [782.430924ms] [201.724622ms] Object stored in database
Trace[501784879]: [3.696487868s] [2.914056944s] Writing response
Trace[501784879]: [4.0511125s] [354.624632ms] END
```

This is related strictly related to what https://github.com/kubernetes/kubernetes/pull/71376 tries to fix. 
We have a large object which is serialized for each watch, given that we have 5K watches it puts a big burden on the master, temporary slowing it down.

I'll try to patch https://github.com/kubernetes/kubernetes/pull/71376 and see if it helps.

**Comment by mm4tt** [comment]
https://github.com/kubernetes/kubernetes/pull/71376 helps significantly. With the same setup (50 nodes, n1-standard-64 master, 5K endpoints watches), creation of a large Endpoints (8K addresses) object took ~600ms. That's almost 8 times faster and what's more important it didn't slow down the whole api server. 

```
I0314 06:57:12.336993       1 trace.go:76] Trace[1168983187]: "Create /api/v1/namespaces/test/endpoints" (started: 2019-03-14 06:57:11.703392589 +0000 UTC m=+46746.763338845) (total time: 633.52889ms):
Trace[1168983187]: [340.699753ms] [340.699753ms] About to convert to expected version
Trace[1168983187]: [511.359644ms] [155.679473ms] Object stored in database
Trace[1168983187]: [633.52889ms] [122.132657ms] END
```

**Comment by wojtek-t** [comment]
That's great new.
I'm waiting for one more experiments (to see how much it helps for memory allocations). But it seems pretty clear we need to prioritize it for 1.15. I will keep that in mind.

**Comment by mm4tt** [comment]
Together with @mborsz we found out that this is causing flakiness in our 5K performance tests. 
Long story short here is what happens:

1. Creating the kube-proxy endpoints object (for prometheus monitoring purposes) slows down api server for ~10s
2. Slow API Server may (or may not, that's why test is flaky) cause some of the controllers to time-out and restart
3. There is one controller (I think it's related to node taints) that, on startup, list all pods on each node without providing resourceVersion. Such requests go to etcd, but there are no indices in etcd which results in iterating over all pods for each node (5K times). The listing takes up to 30 min. If it's restarted and it happens during a test it's very likely that it will increase the api call latency and make the test fail . 

We've disabled prometheus in 5k performance tests in https://github.com/kubernetes/test-infra/pull/11950 to verify it.

Interesting thing is that @mborsz checked the cpu utilization on the machine when the endpoints object was created and it jumped only to about 30-40%. On the other hand we have many evidences that api-server suffered from CPU starvation. It'd suggest that api-server / golang is not able to fully utilize the 64 CPUs on the master. Strange...

**Comment by mm4tt** [comment]
AIs
1. We should reproduce the issue as I did before, but this time also run pprof on the api-server. In addition we should monitor the VM utilization and figure out where the discrepancy comes from.

**Comment by wojtek-t** [comment]
> We should reproduce the issue as I did before, but this time also run pprof on the api-server. In addition we should monitor the VM utilization and figure out where the discrepancy comes from.

pprof may not be the best - I observed that when running pprof, pprof is reporting even 2x+ less than what top is showing. So I would suggest checking both pprof and top.

**Comment by mm4tt** [comment]
I reproduced the issue on the real 5K node cluster.

Creating a large endpoints object resulted in kube-apiserver consuming up to 30 CPUs (max over a couple tries), the VM had still ~50% of CPU resources available. 

```
%Cpu(s): 39.2 us, 12.3 sy,  0.0 ni, 44.7 id,  0.2 wa,  0.0 hi,  3.5 si,  0.0 st
   3578 root      20   0   27.7g  27.5g  71512 S  3060  11.6 332:01.70 kube-apiserver                                                                                                                                                                                                     
   4124 root      20   0 7215652   1.7g 867736 S 308.3   0.7  43:06.67 etcd                                                                                                                                                                                                               
   4096 root      20   0 5523792 345976 132492 S   1.0   0.1   4:22.62 etcd  
``` 

What's worth to notice is that I performed the test when the cluster was in quite stable state. I think the difference between the issue we face in tests is that the large endpoints object is created directly after the cluster is created and when it's still stabilizing. I guess that if the large endpoints was created in that stage we could reach the 100% of CPU for few seconds. If that "cluster stabilizing" stage reports a persistent consumption of 40-50% CPU adding additional 40-50% may result in chocking the VM for a few seconds. 

![XVXebuKRjM7](https://user-images.githubusercontent.com/2604887/55145891-e4839c80-5143-11e9-9f68-27e75eff9a3e.png)

**Comment by mm4tt** [comment]
Sorry, this got unintentionally deprioritized due to too many stuff on my plate, but I'll be getting back to this in Q3. This should get improved with https://github.com/kubernetes/kubernetes/pull/71376

**Comment by wojtek-t** [comment]
This is not exactly what we would like to proceed with.
I'm working on a KEP for that - I will try to get to it early next week: https://github.com/kubernetes/enhancements/pull/924

**Comment by mm4tt** [comment]
Sorry, wasn't clear enough. I was thinking about the idea behind https://github.com/kubernetes/kubernetes/pull/71376, i.e. to not serialize the same object multiple times, not the implementation details of it :)

**Comment by wojtek-t** [comment]
The KEP for it is approved & merged:
https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190329-less-object-serializations.md

Now it's only a matter of execution.
I also opened a small PR (https://github.com/kubernetes/kubernetes/pull/80471) that is adding necessary interface support to apimachinery - all other changes should be very local.

**Comment by wojtek-t** [comment]
FTR: this is the (already closed) last prototype that is closest to the approved proposal: https://github.com/kubernetes/kubernetes/pull/76942
Whoever will be working on it if not me, should look into it and some comments that Daniel already added there.

**Comment by wojtek-t** [comment]
This should be fixed by #81914
